services/api/src/utils/middleware/__tests__/application.js

```diff
@@ -0,0 +1,145 @@
+const { applicationMiddleware } = require('../application');
+const { sleep } = require('../../../utils/sleep');
+const { setupDb, teardownDb, context } = require('../../testing');
+const EventEmitter = require('events');
+
+const { Application, ApplicationEntry } = require('./../../../models');
+const mongoose = require('mongoose');
+
+beforeAll(async () => {
+  await setupDb();
+});
+
+afterAll(async () => {
+  await teardownDb();
+});
+
+describe('application', () => {
+  it('should set an request id', async () => {
+    const application = await Application.create({
+      name: Date.now(),
+      clientId: Date.now(),
+      user: mongoose.Types.ObjectId(),
+    });
+
+    const middleware = applicationMiddleware({ ignorePaths: [] });
+
+    const ctx = context(
+      {
+        headers: {
+          ['client-id']: application.clientId,
+        },
+      },
+      EventEmitter.prototype
+    );
+
+    await middleware(ctx, () => {});
+
+    const [applicationId, hash] = ctx.response.header['request-id'].split('-');
+    expect(applicationId).toBe(application.clientId);
+    expect(hash).toBeDefined();
+    expect(hash).toHaveLength(16);
+  });
+
+  it('should truncate large response data.length < 20', async () => {
+    const application = await Application.create({
+      name: Date.now(),
+      clientId: Date.now(),
+      user: mongoose.Types.ObjectId(),
+    });
+
+    const middleware = applicationMiddleware({ ignorePaths: [] });
+    const ctx = context(
+      {
+        headers: {
+          ['client-id']: application.clientId,
+        },
+      },
+      EventEmitter.prototype
+    );
+
+    ctx.request.body = {
+      password: 'password',
+    };
+
+    ctx.body = {
+      data: [...new Array(23)].map((c, index) => index),
+    };
+
+    await middleware(ctx, () => {});
+    ctx.res.emit('finish');
+    await sleep(50);
```

Comment:
it's due to `res.once('finish', () => {` in the application middleware.
Which again is due to making sure we also capture the output if the request fails

I am not away of simple way to make that synchronously.

Reply:
if the stream has finished though then why the need for sleep?

Reply:
I mean without the `sleep` it WOULD in fact be synchronous

Reply:
as mentioned earlier its due to  https://github.com/bedrockio/bedrock-core/pull/170/files#diff-12b7b46b4ca27ac96733507330408b591495e8a56e4060525653bab0e73bdddcR107 

It's not promised based, and I don't think it can be. As I am trying to make sure that I also capture the output from the error handler.

Reply:
I'm actually doing this in web for env subsitution... same problem:

```js
module.exports = function envMiddleware() {
  const env = JSON.stringify(PUBLIC);
  return async (ctx, next) => {
    await next();
    if (ctx.type === 'text/html' && ctx.body) {
      ctx.body = ctx.body
        .toString()
        .replace(
          '<!--env:conf-->',
          `<script>window.__ENV__ = ${env};</script>`
        );
    }
  };
};
```

Reply:
I want to run after the error handler ... thats the hard part. This will work fine as long as no error is thrown. But i want it to work also when a request fails.

Reply:
ah I see... ok so:

```
try {
  await next();
} finally {
  // ... your stuff
}
```

...no?

Reply:
Nope that wouldn't capture the response from the error handler, as the error handle is last thing to run. 

One option is enforce that Application middleware is the first middleware ... but i don't love it, but that is an option. 

This currently allows for the application middleware to placed anywhere.

Reply:
Right I mean that's the entire middleware pattern right? Middleware isn't designed to be run anywhere, there's an outside -> in flow it has to follow just like `try/catch`  blocks but chainable. We have other middleware that depends on this (error being one of them) ... if applicationMiddleware needs access to the error object then it belongs as the outermost middleware applied. I'd argue that using `res.once` breaks this flow and is what the middleware pattern is trying to avoid.

Reply:
Yeah sure I guess there is some truth to that it, breaks with the middleware conversion. I personally don't think its big issue. 

The end result is same, in terms of functionality ... sure I have todo this little trick to the tests. I gain now have care too much about the order or middleware.

But it's not hill i am willing to die on. Alternative I will have to figure out to detect that the application middleware is the first middleware, and if not throw an error.

Reply:
> But it's not hill i am willing to die on.

me either I mean imho this is the "right" (read "koa") way but you may disagree. I do want to avoid `sleep` in the tests however... what about this:

```js
it('test', async (done) => {

  // ...so some stuff

  await middleware(ctx, () => {});
  ctx.res.once('finish', () => {
    // run assertions here
    done();
  })

});
```

> Alternative I will have to figure out to detect that the application middleware is the first middleware, and if not throw an error.

No, I really wouldn't like to do that. Middleware shouldn't enforce its position in the chain that's not its job. There's plenty of scenarios where another valid middleware could exist outside. Throwing an error is kind of like if you could throw an error if your element isn't topmost z-index stacking order in CSS. It's something that's meant and designed to be overridden - you shouldn't presume anything about context.

Reply:
I can for sure change the test to that.

`Middleware shouldn't enforce its position in the chain that's not its job. ` but this only works correctly if placed correctly. easy to make mistake and not consider the error case and think it's working but it's not.

Reply:
Well you said "detect that the application middleware is the first middleware" which suggests throwing an error based on position, which it shouldn't do. It sounds like instead errorMiddleware is a dependency of applicationMiddleware so maybe it should be removed and wrap the error middleware instead?

Reply:
i have remove the 
```
ctx.res.once('finish', () => {
    // run assertions here
    done();
  })
```

and just enforced that middleware its first + added a comment, as i was doing it, it felt like the right thing todo, specially since only for staging/dev,